[triton] Update pin for PyTorch 2.6/Triton 3.2

bertmaher opened the pull request:
Bump the Triton pin to the release candidate commit for Triton 3.2.

A few changes beyond the pin bump itself are needed:
* Remove the script that adds a git version hash suffix to the Triton wheel, since as of https://github.com/triton-lang/triton/pull/4812 Triton adds that itself
* Add `pybind11` to the Triton build setup, since Triton now depends on it
* Use manylinux-2.28 for the Triton wheel builder, and use clang+lld for building to pick up the right glibc

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

bertmaher committed:
[triton] Update pin for PyTorch 2.6/Triton 3.2

davidberard98 commented:
I'm going to look at the cumsum failure

aakhundov commented:
The following tests should be fixed by #139844:

```
inductor/test_torchinductor.py::GPUTests::test_uint4x2_mixed_mm_cuda
inductor/test_torchinductor.py::GPUTests::test_mixed_mm3_cuda
```

bertmaher committed:
Install pybind11 to build Triton

bertmaher committed:
Revert "Fix triton build failure due to tritonlang.blob.core.windows.net not available (#133694)"

This reverts commit 2cffe82deae5a0a075f817c4c706906f227e3244.

bertmaher committed:
Remove git version suffix from Triton wheel version

It is now always added by Triton's own setup.py (see
https://github.com/triton-lang/triton/pull/4812/)

malfet reviewed (approved):
LGTM, but please mention in PR description which trition change added hash versioning to their builds?

bertmaher commented on .github/scripts/build_triton_wheel.py:
```diff
@@ -81,13 +62,6 @@ def build_triton(
         max_jobs = os.cpu_count() or 1
         env["MAX_JOBS"] = str(max_jobs)
 
-    version_suffix = ""
-    if not release:
-        # Nightly binaries include the triton commit hash, i.e. 2.1.0+e6216047b8
-        # while release build should only include the version, i.e. 2.1.0
-        version_suffix = f"+{commit_hash[:10]}"
-        version += version_suffix
-
```
I meant to put this in the summary but it's this PR that now adds a version suffix to all Triton builds: https://github.com/triton-lang/triton/pull/4812.  So with our version-suffixing logic we end up with things like `3.2.0+git1234abcd+1234abcd`, which is an invalid version string

jataylo commented:
When these two are merged it should address majority of ROCm failures.
https://github.com/pytorch/pytorch/pull/139881
https://github.com/pytorch/pytorch/pull/139883

Lets rebase and run UTs again once we get these in and we can reassess any new failures.

jataylo commented:
Note https://github.com/triton-lang/triton/pull/5089 is known to resolve a pretty big regression on AMD backend, will be nice to move pin after this at some point, or cherry pick.

bertmaher commented:
> Note [triton-lang/triton#5089](https://github.com/triton-lang/triton/pull/5089) is known to resolve a pretty big regression on AMD backend, will be nice to move pin after this at some point, or cherry pick.

@jataylo Do you think #5089 will cherry-pick cleanly onto https://github.com/triton-lang/triton/tree/rc/3.2.x?  It's logistically easier for me to do cherry-picks than a pin bump (mostly for annoying internal-to-meta reasons, but still)

jataylo commented:
> > Note [triton-lang/triton#5089](https://github.com/triton-lang/triton/pull/5089) is known to resolve a pretty big regression on AMD backend, will be nice to move pin after this at some point, or cherry pick.
> 
> @jataylo Do you think #5089 will cherry-pick cleanly onto https://github.com/triton-lang/triton/tree/rc/3.2.x? It's logistically easier for me to do cherry-picks than a pin bump (mostly for annoying internal-to-meta reasons, but still)

Hopefully shouldn't be too bad, lets see. So we are now at a stable pin point? (assuming there isn't some big issue)

jataylo commented:
ROCm tests are almost green, we're failing flex attention unit tests due to OOMing on shard memory still. https://github.com/pytorch/pytorch/pull/139883 was supposed to resolve this but there might have been an edge case I've missed out.

atalman committed:
Update build-triton-wheel.yml - use manylinux_2_28-builder

atalman committed:
Use manylinux2_28-builder

atalman committed:
Remove gcc 11 install fro xpu builds

atalman committed:
Update build-triton-wheel.yml

atalman committed:
Add TRITON_BUILD_WITH_CLANG_LLD

jataylo commented:
Flex attention failures on ROCm to be resolved when https://github.com/pytorch/pytorch/pull/140270 is finished merging.

bertmaher committed:
Add pybind11 to conda build

bertmaher committed:
Only call dnf on the cuda runners

bertmaher commented:
@chuanqi129 @Stonepia @etaf The Triton XPU wheels are failing to build, because the upstream Triton version # is bumping to 3.2.  How would you like to proceed?

etaf commented:
duplicate with https://github.com/pytorch/pytorch/pull/137886

chuanqi129 commented:
> @chuanqi129 @Stonepia @etaf The Triton XPU wheels are failing to build, because the upstream Triton version # is bumping to 3.2. How would you like to proceed?

@bertmaher Yes, we also want to do it. Could you please add the PR https://github.com/pytorch/pytorch/pull/137886 change into your PR, so that we can update together

bertmaher commented:
> @bertmaher Yes, we also want to do it by PR #137886. Seems we need to land this PR with xpu triton build failure to bump up the triton version firstly. And then land the #137886. How do you think?

That sounds fine to me!  I'll go ahead and land this one once the remaining non-xpu failures are sorted out

bertmaher committed:
Only use clang/lld for CUDA

bertmaher committed:
Bump pin to include ProgramPoint fix (triton-lang/triton#5181)

jataylo commented:
Hey @bertmaher any chance we can rebase to stop the docker build timeouts on ROCm jobs? we _should_ be green now based on my testing PR (https://github.com/pytorch/pytorch/pull/140698)

bertmaher commented:
Pushed a rebase this morning.  There are a few expected failures left:
- the `test_convolution_as_mm` failure, which has been intermittent, but is now confirmed to be real.  @davidberard98, @embg, and I are teaming up on this one
- the xpu wheel builds complaining about the 3.2.0 version string is going to be fixed by the xpu team
- the failure to satisfy version requirements failure will only go away after this diff lands, because we bumped the version so the new version hasn't been uploaded.  (I guess there's fundamentally no way to solve this, unfortunately)

jataylo commented:
Still seeing ROCm build timeouts https://github.com/pytorch/pytorch/pull/140851 will resolve when merged...

bertmaher committed:
Bump to include user-defined autotuner fix and load-with-other codegen workaround

bertmaher commented:
@pytorchbot merge -f "Remaining failures are either expected (xpu, and Triton wheel version) or due to self-hosted runner timeouts"

jataylo commented:
@bertmaher this merge may break rocm as the docker build jobs are timing out until https://github.com/pytorch/pytorch/pull/140851 is merged. Lets see

bertmaher commented:
@chuanqi129 @etaf This version bump has landed, if you'd like to go ahead with bumping the xpu triton version, please proceed!

atalman commented:
@pytorchmergebot revert -c weird -m "Will revert and reland skipping xpu builds"

bertmaher committed:
Remove xpu temporarily until its triton version is fixed

atalman commented:
@pytorchmergebot merge -f "reland - lint is green"

guangyey commented:
Thanks @bertmaher, we will rebase our PR https://github.com/pytorch/pytorch/pull/137886 to update XPU triton.

cpuhrsch commented:
This seems to break int_mm https://github.com/pytorch/pytorch/issues/144705 . I'l try to produce a standalone Triton kernel to reproduce this error.
